MIPs/mip-3.md

```diff
@@ -0,0 +1,186 @@
+Author(s): Tomás Santos (@tommasini)
+Type: Community
+Created: 2023-11-10
+---
+
+## Summary
+
+This proposal introduces a new JSON RPC method `wallet_swapAsset` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.
+
+## Motivation
+
+This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.
+
+Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.
+
+Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:
+
+- Improved usability - Users can get the right tokens for dApp operations in fewer steps.
+- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
+- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
+- Interoperability - Standard method works across different wallets.
+- Increase revenue - Easier adoption means more users will use MetaMask swaps.
+
+Comparing with other alternative, implementing it using snaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The goal is to reduce friction to swaps.
+
+Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.
+
+# Usage Example
+
+```
+ethereum.request({
+  method: 'wallet_swapAsset',
+  params: [{
+    from: [{
+      tokenAddress: '0x1234567890abcdefABCDEF1234567890ABCDEF',
+      value: '0xDE0B6B3A7640000',
+    }],
+    to: {
+      tokenAddress: '0xabcdef1234567890ABCDEF1234567890abcdef',
+    },
+    userAddress: '0x0000000000000000000000000000000000000000'
+  },
+  ]
+});
+
+```
+
+## Proposal Specification
+
+The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:
+
+- `from`: An object containing details about the source token. It should include:
+
+  - `tokenAddress`: The address of the source token.
+  - `value`: The amount of wei in hexadecimal format of the source token to be swapped.
+
+- `to`: An object containing details about the destination token. It should include:
+
+  - `tokenAddress`: The address of the destination token.
+
+- `userAddress`: Account address connected to the dapp.
+
+- `referralCode`: (Optional) Future implementation will allow adding a refferal code to drive adoption of the JSON RPC method across the builders community.
```

Comment:
```suggestion
- `referralCode`: (Optional) Future enhancements to this API will allow adding a referral code by which referrers can earn a percent of swap fees associated with the swaps they propose. This will help to further drive adoption of this API.
```